### PR TITLE
[SMTChecker] Fix handling of assignments of array/mapping state variable accessed using contract name.

### LINKED ISSUE #10831: [SMTChecker] ICE in member access because symbStruct is null
## Description

```
pragma experimental SMTChecker;
contract A {
  int[] a;
  function f() public {
    (A.a[a.length]=2);
  }
}
```

throws

https://github.com/ethereum/solidity/blob/a75b87c80eae69b4fc0916000593ce4a3614f061/libsolidity/formal/SMTEncoder.cpp#L1504

### LINKED ISSUE #10847: [SMTChecker] False negative with assignment to state variables of reference type
In the following example:
```
pragma experimental SMTChecker;
contract A {
	int[] a;
	function f() public {
		require(a.length == 1 && a[0] == 1);
		int[] storage u = a;
		assert(u[0] == 1); // should hold
		int[] memory b = new int[](2);
		a = b;
		assert(u[0] == 1); // should fail
		A.a = b;
		assert(u[0] == 1); // should fail
	}

	function push_v(int x) public {
		a.push(x);
	}
}
```
SMTChecker currently reports assertion violation for the third assertion, but not the second one. The second assertion should also be reported as violated.

The problem is slightly different handling of assignments based on whether or not the contract type is used.

### PR DESCRIPTION
**Pull request by blishko**
Fixes #10831.

### CODE REVIEW

File: libsolidity/formal/SMTEncoder.cpp [hunk 1]
```diff
@@ -2158,30 +2129,33 @@ void SMTEncoder::expressionToTupleAssignment(vector<shared_ptr<VariableDeclarati
...
 	// even if there is implicit conversion.
 	// This is a special case where the SMT sorts are different.
 	// For now we are unaware of other cases where this happens, but if they do appear
 	// we should extract this into an `implicitConversion` function.
-	assignment(_variable, expr(_value, _variable.type()));
+	assignment(_variable, expr(_value, _variable.type()), true);
 }
 
-void SMTEncoder::assignment(VariableDeclaration const& _variable, smtutil::Expression const& _value)
+void SMTEncoder::assignment(VariableDeclaration const& _variable, smtutil::Expression const& _value, bool initialization)
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
void SMTEncoder::assignment(VariableDeclaration const& _variable, smtutil::Expression const& _value, bool _initialization)
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 2]
```diff
@@ -2680,6 +2654,14 @@ VariableDeclaration const* SMTEncoder::identifierToVariable(Expression const& _e
 			solAssert(m_context.knownVariable(*varDecl), "");
 			return varDecl;
 		}
+	// But we are interested in "contract.var", because that is the same as just "var".
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:2675]: Isn't this case already handled by `memberAccess`?

File: libsolidity/formal/SMTEncoder.cpp [hunk 3]
```diff
@@ -1653,22 +1623,7 @@ void SMTEncoder::arrayPushPopAssign(Expression const& _expr, smtutil::Expression
...
-			resetReferences(*varDecl);
-		m_context.addAssertion(m_context.newValue(*varDecl) == _array);
-		m_context.expression(*id)->increaseIndex();
-		defineExpr(*id,currentValue(*varDecl));
-	}
-	else if (
-		dynamic_cast<IndexAccess const*>(expr) ||
-		dynamic_cast<MemberAccess const*>(expr)
-	)
-		indexOrMemberAssignment(_expr, _array);
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: Why is it not necessary anymore to call this function?

File: libsolidity/formal/SMTEncoder.cpp [hunk 4]
```diff
@@ -523,35 +524,20 @@ void SMTEncoder::endVisit(UnaryOperation const& _op)
 				_op
 			).first;
 		};
-		if (auto identifier = dynamic_cast<Identifier const*>(subExpr))
-		{
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: Where did this block go?

File: libsolidity/formal/SMTEncoder.cpp [hunk 5]
```diff
@@ -523,35 +524,20 @@ void SMTEncoder::endVisit(UnaryOperation const& _op)
...
+			arrayPushPopAssign(*subExpr, newValue);
 		}
-		else if (isEmptyPush(*subExpr))
+		else
 		{
 			auto innerValue = expr(*subExpr);
 			auto newValue = computeNewValue(innerValue);
 			defineExpr(_op, _op.isPrefixOperation() ? newValue : innerValue);
-			arrayPushPopAssign(*subExpr, newValue);
+			assignment(*subExpr, newValue, subExpr->annotation().type);
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: Ah I guess the block that I asked "where did this go" ends up here instead?

**Review by leonardoalt — COMMENTED** [review]
A few comments/questions, but we can also have a call to discuss it.

**Comment by blishko** [comment]
@leonardoalt , after our discussion, I decided to split the original PR into smaller parts. This one is the first.

File: libsolidity/formal/SMTEncoder.cpp (hunk 2 above)
Comment (blishko) [reply libsolidity/formal/SMTEncoder.cpp:2675]: As discussed, `visit` for `memberAccess` handles the creation of SMT expression for such solidity Expression, but in here I need to get to the state variable's `VariableDeclaration`.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**